src/VisualStudio/Core/Def/Implementation/LanguageClient/LanguageServerClientEventListener.cs

```diff
@@ -13,19 +19,57 @@ namespace Microsoft.VisualStudio.LanguageServices.Implementation.LanguageService
     internal class LanguageServerClientEventListener : IEventListener<object>
     {
+        private readonly LanguageServerClient _languageServerClient;
+        private readonly Lazy<ILanguageClientBroker> _languageClientBroker;
         private readonly TaskCompletionSource<object> _taskCompletionSource;
 
         public Task WorkspaceStarted => _taskCompletionSource.Task;
 
-        public LanguageServerClientEventListener()
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public LanguageServerClientEventListener(LanguageServerClient languageServerClient, Lazy<ILanguageClientBroker> languageClientBroker)
         {
-            _taskCompletionSource = new TaskCompletionSource<object>();
+            this._languageServerClient = languageServerClient;
+            this._languageClientBroker = languageClientBroker;
+            this._taskCompletionSource = new TaskCompletionSource<object>();
         }
 
+        /// <summary>
+        /// LSP clients do not necessarily know which language servers (and when) to activate as they are language agnostic.
+        /// We know we can provide <see cref="LanguageServerClient"/> as soon as the workspace is started, so tell the
+        /// <see cref="ILanguageClientBroker"/> to start loading it.
+        /// </summary>
         public void StartListening(Workspace workspace, object serviceOpt)
         {
             // mark that roslyn solution is added
-            _taskCompletionSource.SetResult(null);
+            this._taskCompletionSource.SetResult(null);
+
+            this.WorkspaceStarted.ContinueWith(_ =>
+            {
+                // Trigger a fire and forget request to the VS LSP client to load our ILanguageClient.
+                // This needs to be done with .Forget() as the LoadAsync (VS LSP client) synchronously stores the result task of OnLoadedAsync.
+                // The synchronous execution happens under the sln load threaded wait dialog, so LSP requests cannot be made in between triggering LoadAsync and storing the result task from OnLoadedAsync.
+                // The result task from OnLoadedAsync is waited on before invoking LSP requests to the ILanguageClient.
+                this._languageClientBroker.Value.LoadAsync(new LanguageClientMetadata(new string[] { ContentTypeNames.CSharpContentType, ContentTypeNames.VisualBasicContentType }), this._languageServerClient).Forget();
+            }, TaskScheduler.Default);
+        }
+
+        /// <summary>
+        /// The <see cref="ILanguageClientBroker.LoadAsync(ILanguageClientMetadata, ILanguageClient)"/> 
+        /// requires that we pass the <see cref="ILanguageClientMetadata"/> along with the language client instance.
+        /// The implementation of <see cref="ILanguageClientMetadata"/> is not public, so have to re-implement.
+        /// </summary>
+        private class LanguageClientMetadata : ILanguageClientMetadata
```

Comment:
This is unfortunate, we should probably file a bug to have an implementation of this in the LSP client definitions.

Reply:
@tinaschrepfer to comment here - it also seems like we shouldn't need to pass along the metadata if we have the actual client instance?  Because as a result we're keeping this metadata in two places that must be kept in sync

Reply:
This does seem like a deficiency.  We should either have the class available, or allow you to pass an instance of ILanguageClient and we learn the parse the metadata ourselves.  Please file a bug.

Reply:
Done
https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1043922